pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/TransactionCursorImpl.java

```diff
@@ -123,4 +204,377 @@ private void addTxnToCommittedIndex(TxnID txnID, long committedAtLedgerId) {
+                            case START:
+                                recoverFromStart(currentPosition).whenComplete((ignore, error) -> {
+                                    checkComplete(error, recoverFuture);
+                                });
+                                break;
+                            case MIDDLE:
+                                recoverFromMiddle(snapshot).whenComplete((ignore, error) -> {
+                                    checkComplete(error, recoverFuture);
+                                });
+                                break;
+                            case END:
+                                recoverFromEnd(snapshot).whenComplete((ignore, error) -> {
+                                    checkComplete(error, recoverFuture);
+                                });
+                        }
+                    }
+                }
+            });
+
+        return recoverFuture;
+    }
+
+    private void checkComplete(Throwable error, CompletableFuture<Void> future) {
+        if (error != null) {
+            future.completeExceptionally(error);
+        } else {
+            future.complete(null);
+        }
+    }
+
+    @Getter
+    final static class PersistentTxnIndexSnapshot {
+        enum SnapshotStatus  {
+            START,
+            MIDDLE,
+            END,
+        }
+
+        SnapshotStatus status;
+        // If the status is START, the position is the position which is the transaction log doing snapshot.
+        //  If the status is others, the position is the snapshot beginning position on the cursor ledger.
+        Position position;
+        TransactionMetaImpl meta;
+
+        PersistentTxnIndexSnapshot(byte[] entry) {
+            StoredTxn txn = DataFormat.parseStoredTxn(entry);
+            switch (txn.getStoredStatus()) {
+                case START:
+                    this.status = SnapshotStatus.START;
+                    break;
+                case MIDDLE:
+                    this.status = SnapshotStatus.MIDDLE;
+                    break;
+                case END:
+                    this.status = SnapshotStatus.END;
+                    break;
+            }
+            this.meta = DataFormat.recoverMeta(txn.getTxnMeta());
+            this.position = DataFormat.recoverPosition(txn.getPosition());
+        }
+
+    }
+
+    private CompletableFuture<Void> recoverFromStart(Position currentPosition) {
+        return readPrevEntry(currentPosition)
+                   .thenApply(entry -> new PersistentTxnIndexSnapshot(entry.getData()))
+                   .thenCompose(this::recoverFromEnd);
+
+    }
+
+    private CompletableFuture<Entry> readPrevEntry(Position position) {
+        PositionImpl currentPos = (PositionImpl) position;
+        if (currentPos.getEntryId() == 0) {
+            return FutureUtil.failedFuture(
+                new TransactionIndexRecoveringError("Not found the prev position of the current position " + position));
```

Comment:
when does this happen?

if the snapshot is not written completely, you can still replay the index from the transaction buffer, no? does your logic handle `TransactionIndexRecoveringError`?

Reply:
> when does this happen?

When reading entries on the cursor ledger, if the entry id is 0, it indicates there is no more entry before the read position.